src/prompts/password.rs

```diff
@@ -342,30 +388,61 @@ impl<'a> PasswordPrompt<'a> {
     }
 
     fn prompt<B: PasswordBackend>(mut self, backend: &mut B) -> InquireResult<String> {
-        let final_answer: String;
-
-        loop {
+        let final_answer = loop {
             self.render(backend)?;
 
             let key = backend.read_key()?;
 
             match key {
                 Key::Interrupt => interrupt_prompt!(),
-                Key::Cancel => cancel_prompt!(backend, self.message),
+                Key::Cancel if self.confirmation_first_answer.is_none() => {
+                    cancel_prompt!(backend, self.main_message)
+                }
+                Key::Cancel => {
+                    self.error = None;
+                    self.input.clear();
+                    self.confirmation_first_answer.take();
+                    self.message = Cow::Borrowed(self.main_message);
+                }
                 Key::Submit => match self.validate_current_answer()? {
+                    Validation::Valid if !self.enable_confirmation => break self.cur_answer(),
                     Validation::Valid => {
-                        final_answer = self.cur_answer();
-                        break;
+                        match self
+                            .confirmation_first_answer
+                            .take()
+                            .as_ref()
+                            .map(Input::content)
+                        {
+                            None => {
+                                self.error = None;
+                                self.confirmation_first_answer =
```

Comment:
I created a new field `confirmation_stage` which keeps track of if we are in the confirmation stage or not. I initially encoded the `input` field in `PasswordConfirmation` as an `Option`, as it would be unambiguous as to if the input is usable or not. Eventually, I reverted back to a plain `Input` and a flag, for optimization reasons and a slightly more convenient flow with the ESC flow for the confirmation prompt (see upcoming comment about this).